src/Shared/DiagnosticSourceInstrumentation/PropertyFetcher.cs

```diff
@@ -85,54 +73,109 @@ public bool TryFetch(object obj, out T value, bool skipObjNullCheck = false)
-                var instantiatedTypedPropertyFetcher = typedPropertyFetcher.MakeGenericType(
-                    typeof(T), propertyInfo.DeclaringType, propertyInfo.PropertyType);
-                return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, propertyInfo);
+                var declaringType = propertyInfo.DeclaringType;
+                if (declaringType!.IsValueType)
+                {
+                    throw new NotSupportedException(
+                        $"Type: {declaringType.FullName} is a value type. PropertyFetcher can only operate on reference payload types.");
+                }
+
+                if (declaringType == typeof(object))
+                {
+                    // This is only necessary on .NET 7.
+                    // It is used to force the AOT compiler to create an instantiation of the method with a reference type.
+                    // The code for that instantiation can then be reused at runtime to create instantiation over any other reference.
+                    // TODO: when adding net8.0 as the project target:
+                    // This is not needed in .NET 8, because the compiler is improved and call into MakeGenericMethod will be AOT-compatible.
+                    return CreateInstantiated<object>(propertyInfo);
+                }
+                else
+                {
+                    return DynamicInstantiationHelper(declaringType, propertyInfo);
+                }
+
+                // Separated as a local function to be able to target the suppression to just this call.
+                // IL3050 was generated here because of the call to MakeGenericType, which is problematic in AOT if one of the type parameters is a value type;
+                // because the compiler might need to generate code specific to that type.
+                // If the type parameter is a reference type, there will be no problem; because the generated code can be shared among all reference type instantiations.
+#if NET6_0_OR_GREATER
+                [RequiresUnreferencedCode(TrimCompatibilityMessage)]
+                [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "The code guarantees that all the generic parameters are reference types.")]
+#endif
+                static PropertyFetch? DynamicInstantiationHelper(Type declaringType, PropertyInfo propertyInfo)
+                {
+                    return (PropertyFetch?)typeof(PropertyFetch)
+                        .GetMethod(nameof(CreateInstantiated), BindingFlags.NonPublic | BindingFlags.Static)!
+                        .MakeGenericMethod(declaringType) // This is validated in the earlier call chain to be a reference type.
+                        .Invoke(null, new object[] { propertyInfo })!;
+                }
             }
         }
 
-        public virtual bool TryFetch(object obj, out T value)
+        public virtual bool TryFetch(object obj, out T? value)
```

Comment:
Since we don't ever create instances of the concrete `PropertyFetch` class, we could just make the class and this method `abstract`.

```suggestion
        public abstract bool TryFetch(object obj, out T? value);
```